### PR TITLE
Add UI for editing channel types

### LINKED ISSUE #830: Channel Settings: channel type
The top half of: https://invis.io/XSLE63T6GR2#/303979450_Settings__Channel

### acceptance criteria:
- [ ] "Basic" tab
- [ ] choose public, restricted or private channel
- [ ] Any moderator can change the channel settings

### PR DESCRIPTION
**Pull request by noisecapella**
#### What are the relevant tickets?
Fixes #830 

#### What's this PR do?
Refactors the UI to look a little more similar to the mockup*. Updates backend to allow restricted type.

*See: https://projects.invisionapp.com/share/XSLE63T6GR2#/screens/303979450

#### How should this be manually tested?
As a moderator verify that you can change the channel type. You should also be able to edit the description in the appearance tab.

If you click Cancel then click the edit icon again the text should revert to what it was before.

### CODE REVIEW

**Comment by aliceriot** [comment]
It looks in the mockups like the 'save' and 'cancel' buttons should be outside of the card, not sure how much we care though since the designs are probably not yet final:

mockup_buttons

File: static/js/containers/admin/EditChannelBasicPage.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+  const processing = state.channels.processing
+  return {
+    channel,
+    channelName,
+    processing,
+    channelForm: getForm(state.forms)
+  }
+}
+
+export default connect(mapStateToProps)(EditChannelBasicPage)
```
[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/containers/admin/EditChannelBasicPage.js]: here I think if you use the `withSingleColumn` HOC you should be able to do something like

```js
render () {
  return <React.Fragment>
    <DocumentTitle />
    <Form />
  </React.Fragment>
}
```

and that should take care of the layout. If you want to go that direction, I think there's an extraneous wrapping div in the `AdminPage` container which would also need to be turned into a `React.Fragment`

**Comment by noisecapella** [comment]
Regarding save/cancel I talked to @pdpinch and I'm going to leave it as is, at least for this PR

File: static/js/containers/admin/AdminPage.js [hunk 2]
```diff
@@ -21,8 +22,12 @@ export default class AdminPage extends React.Component<*, void> {
...
         />
         <Route
-          path={`${match.url}/channel/edit/:channelName`}
-          component={EditChannelPage}
+          path={`${match.url}/channel/edit/basic/:channelName`}
+          component={EditChannelBasicPage}
+        />
+        <Route
+          path={`${match.url}/channel/edit/appearance/:channelName`}
+          component={EditChannelAppearancePage}
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/containers/admin/AdminPage.js]: Sounds good

[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/containers/admin/AdminPage.js]: Do you remember what the react-router v4 equivalent is for `IndexRedirect`?

File: static/js/containers/admin/EditChannelBasicPage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/containers/admin/EditChannelBasicPage.js]: Oh that's nice

File: static/js/containers/admin/AdminPage.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/containers/admin/AdminPage.js]: bah I always forget they changed everything, they might have removed that functionality. we could I think emulate it by having a `<Route />` for the route we want which just rendered a component which returned a `<Redirect />`, but that's a little silly, so if you want we can just not, haha

File: static/js/containers/admin/EditChannelBasicPage.js [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+          ...EDIT_CHANNEL_PAYLOAD,
+          errors: validation.value
+        })
+      )
+    } else {
+      dispatch(actions.channels.patch(channelForm.value)).then(channel => {
+        history.push(channelURL(channel.name))
+      })
+    }
+  }
```
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/containers/admin/EditChannelBasicPage.js]: I talked with @rhysyngsun and he said it doesn't currently support custom `onUpdate` logic which I need for the next PR #852

**Comment by noisecapella** [comment]
Ok I think that's everything

File: static/js/containers/admin/EditChannelBasicPage.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/containers/admin/EditChannelBasicPage.js]: ah, that's too bad :/

File: static/js/containers/admin/EditChannelBasicPage.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (aliceriot) [reply static/js/containers/admin/EditChannelBasicPage.js]: I am a _huge_ fan of `<React.Fragment>`, haha

**Comment by aliceriot** [comment]
cool, will take a look in a sec! It looks like something it failing on travis, not sure what's up

File: static/js/containers/admin/AdminPage.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (noisecapella) [reply static/js/containers/admin/AdminPage.js]: I'll leave it as is, I'm not sure we have a use case for this route at the moment

**Comment by noisecapella** [comment]
@rhysyngsun said that we are using an old version of flow, so I'm looking at upgrading now

**Comment by aliceriot** [comment]
I've been able to get it to pass by doing `import React, { Fragment } from 'react'`

**Comment by aliceriot** [comment]
(if that's the thing it's erroring on)

**Comment by noisecapella** [comment]
I managed to get it working with a few flow tweaks here and there, let me know what you think

**Comment by aliceriot** [comment]
cool, new version of flow here we come!

**Comment by noisecapella** [comment]
By the way what's the significance of the flow version in `.eslintrc`?

**Comment by aliceriot** [comment]
I think it tells it to ignore some rules which are only relevant to later versions

**Comment by aliceriot** [comment]
Same thing for specifying the react version in there, which is kind of annoying to have to do

**Pull request merged**